### ISSUE TITLE
Connection closes when starting a upload (or download) while one is active

### ISSUE THREAD

**Issue by gusnan** [masneyb/gftp#128]
gftp 2.8.0b, on Debian stable, Xfce.

As the title, after starting a few uploads (marking several files, and starting the download, marking overwrite), and while those are in progress, starting another, the connection to the remote server is lost.

This is Debian https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=726880 , while they talk about downloads, I have tested with uploads, and can reproduce the behavior.

**Comment by wdlkmpx** [comment]
Yeah this is probably the most outstanding bug. Hopefully it will be fixed for 3.0. This is related to #121

Before 2.8.0b, I think gftp would crash if you tried to do that and other operations while a file transfer is ongoing. It was a side effect of the transition from GtkCList -> GtkTreeView.

So 2.8.0b is ready to replace all previous gftp versions, it's still buggy and prone to voodoo, but solves several crashes and bugs that make **ALL** older versions unusable. It should compile and work fine even in Debian Squeeze (2011). 

I think this bug is also fixed
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=582354

Don't hesitate to open as many issues as needed. I just found that `gFTP -> Preferences -> General -> Connect to remote server on startup` is broken. It's just a matter of testing the features/options.. is it broken? The answer can only be yes or no. Should I remove broken options? The answer can only be yes or no.

**Comment by wdlkmpx** [comment]
In my quest for the truth I found this:
(https://salsa.debian.org/debian/gftp/-/commit/c73fbfbc36ae3a1418a886b2b15d106945c18522)

Is there a problem with the openssl support? An Arch Linux user said it works fine with implicit TLS (see issue 127), which is just a hack to the standard TLS support fixed and improved by RocketMan

2.8.0b supports only explicit TLS (unencrypted connection + AUTH TLS to activate the encrypted channelz), implicit TLS will be available in 2.9.0b along with MLSD and other important changes ...

I tested it on Debian Stretch with vsftpd, pure-FTPd and a Windows FTP test server provided by Revex , so it basically passed all the tests in this somewhat old distro.

**Comment by gusnan** [comment]
It's not a problem with the functionality, I got a bit worried about the licensing, since I found this:

https://lists.debian.org/debian-legal/2018/12/msg00018.html

which claims that GPLv2 cannot link to openssl, since it's under the apache2 license.

Also see
https://www.apache.org/licenses/GPL-compatibility.html

I don't want to rush it into Debian, and get into licensing problems. However, all is not dark, to me it looks like openssl3 (which currently is in Debian experimental) would make the problems go away. Is there any progress in getting gftp working with openSSL 3.0?

If I am understanding things wrong, please educate me.

**Comment by wdlkmpx** [comment]
The Apache 2.0 license is compatible with GPL3 but not with GPL2. Funny many people consider GPL2 the last acceptable GPL license. 

The other option is GnuTLS, but GFTP is barely alive, it's a miracle development is still happening, so OpenSSL support is the only option

As someone who has edited extensive portions of the code with a (ongoing) major rewrite in the GTK UI code, I suggest changing the license to MIT 
https://opensource.org/licenses/MIT

Or ISC
https://opensource.org/licenses/ISC

**Comment by MartinX3** [comment]
I can reproduce it with 2.8.0b
If I change the directory while transfering a file, I loose the connection.
```
listbox.c: ERROR, could not retrieve filename(s)...
* on_listbox_row_activated_cb(): ERROR
```

If I then do a right click on the transfer and click on "remove file", I get the exception
`** (gftp-gtk:384786): CRITICAL **: 21:41:16.001: gftp_fd_write: assertion 'fd >= 0' failed`
